src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs

```diff
@@ -0,0 +1,365 @@
+            }
+
+            _outputReferences?.Execute(context);
+
+            return _reduceDocuments.Count;
+        }
+
+        public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+        {
+            return new OutputReduceToCollectionCommandDto
+            {
+                OutputReduceToCollection = _outputReduceToCollection, ReduceOutputIndex = _reduceOutputIndex, ReduceDocuments = _reduceDocuments
+            };
+        }
+
+        public void AddReduce(string reduceKeyHash, object reduceObject)
+        {
+            if (_reduceDocuments.TryGetValue(reduceKeyHash, out var outputs) == false)
+            {
+                outputs = new List<(BlittableJsonReaderObject Json, string ReferenceDocumentId)>(1);
+                _reduceDocuments.Add(reduceKeyHash, outputs);
+            }
+
+            var djv = new DynamicJsonValue();
+
+            if (_index.OutputReduceToCollectionPropertyAccessor == null)
+                _index.OutputReduceToCollectionPropertyAccessor = PropertyAccessor.Create(reduceObject.GetType(), reduceObject);
+
+            OutputReferencesPattern.DocumentIdBuilder referenceDocIdBuilder = null;
+
+            string referenceDocumentId = null;
+
+            using (_patternForReduceOutputReferences?.BuildReferenceDocumentId(out referenceDocIdBuilder))
+            {
+                foreach (var property in _index.OutputReduceToCollectionPropertyAccessor.GetPropertiesInOrder(reduceObject))
+                {
+                    var value = property.Value;
+                    djv[property.Key] = TypeConverter.ToBlittableSupportedType(value, context: _indexContext);
+
+                    if (referenceDocIdBuilder?.ContainsField(property.Key) == true)
+                        referenceDocIdBuilder.Add(property.Key, property.Value);
+                }
+
+                if (referenceDocIdBuilder != null)
+                    referenceDocumentId = referenceDocIdBuilder.GetId();
+            }
+
+            djv[Constants.Documents.Metadata.Key] = new DynamicJsonValue
+            {
+                [Constants.Documents.Metadata.Collection] = _outputReduceToCollection
+            };
+
+            var doc = _indexContext.ReadObject(djv, "output-of-reduce-doc", BlittableJsonDocumentBuilder.UsageMode.ToDisk);
+
+            outputs.Add((doc, referenceDocumentId));
+        }
+
+        public void DeleteReduce(string reduceKeyHash)
+        {
+            _reduceKeyHashesToDelete.Add(reduceKeyHash);
+
+            if (_reduceDocuments.Remove(reduceKeyHash, out var outputs))
+            {
+                foreach (var item in outputs)
+                {
+                    item.Json.Dispose();
+                }
+            }
+        }
+
+        public static bool IsOutputDocumentPrefix(string prefix)
+        {
+            var parts = prefix.Split("/", StringSplitOptions.RemoveEmptyEntries);
```

Comment:
Can we try this without allocations?

Reply:
We call it very occasionally - once per indexing batch. For validation purposes only:

https://github.com/ravendb/ravendb/blob/9ce10e6847103dafa0f59c75ab5097a4b741b28d/src/Raven.Server/Documents/Indexes/Persistence/DeleteReduceOutputDocumentsCommand.cs#L16-L17

I don't think it's worth to optimize that